File: Framework/RadialBasisFunction/rbfKMeans.cpp
//
// Created by juanm on 03/07/2022.
//

#include <random>
#include <iostream>
#include <iomanip>
#include <string>

#include "Eigen/Dense"

using namespace Eigen;
using namespace std;

auto asMatrix(float **array, int numRows, int numCols) {
    MatrixXf toMatrix;
    toMatrix.resize(numRows, numCols);
    int i = 0;
    do {
        for (int j = 0; j < numCols; j += 1) {
            toMatrix(i, j) = array[i][j];
        }
        i += 1;
    } while (i < numRows);
    return toMatrix;
}

auto **asArray(Matrix<float, -1, -1> matrixInput, int numRows, int numCols) {
    auto **toArray = new float *[numRows];
    int i = 0;
    do {
        toArray[i] = new float[numCols];
        for (int j = 0; j < numCols; j += 1) {
            toArray[i][j] = matrixInput(i, j);
        }
        i += 1;
    } while (i < numRows);
    return toArray;
}

auto rowMatrix(Matrix<float, -1, -1> matrixToConvert, int numRow) {
    MatrixXf row;
    row.resize(1, matrixToConvert.cols());

    for (int i = 0; i < matrixToConvert.cols(); i += 1) {
        row(0, i) = matrixToConvert(numRow, i);
    }
    return row;
}

tuple<int, int> numRowsAndCols(Matrix<float, -1, -1> matrixInput) {
    int rows = matrixInput.rows();
    int cols = matrixInput.cols();
    return {rows, cols};
}

Matrix<float, -1, -1> flat(Matrix<float, -1, -1> matrixToFlat) {
    int size = matrixToFlat.rows() * matrixToFlat.cols();

    MatrixXf flatten;
    flatten.resize(1, size);

    int counter = 0;

    for (int i = 0; i < matrixToFlat.rows(); i += 1) {
        for (int j = 0; j < matrixToFlat.cols(); j += 1) {
            flatten(0, counter) = matrixToFlat(i, j);
            if (counter < size) counter += 1;
        }
    }
    return flatten;
}

float linalg(Matrix<float, -1, -1> matrixInput) {
    float total = 0.;

    for (int i = 0; i < matrixInput.rows(); i += 1) {
        for (int j = 0; j < matrixInput.cols(); j += 1) {
            total = total + pow(abs(matrixInput(i, j)), 2);
        }
    }
    total = pow(total, 0.5);
    return total;
}

int intRandomizer(int lower_bound, int upper_bound) {
    srand(time(NULL));
    return rand() % (upper_bound - lower_bound + 1) + lower_bound;
}

// Switch values in matrix
auto shuffle(Matrix<float, -1, -1> inputMatrix) {
    int randRow1 = 0;
    int randCol1 = 0;
    int randRow2 = 0;
    int randCol2 = 0;

    for (int i = 0; i < inputMatrix.rows(); i += 1) {
        do {
            randRow1 = intRandomizer(0, inputMatrix.rows() - 1);
            randCol1 = intRandomizer(0, inputMatrix.cols() - 1);
            randRow2 = intRandomizer(0, inputMatrix.rows() - 1);
            randCol2 = intRandomizer(0, inputMatrix.cols() - 1);
        } while (randRow1 == randRow2 && randCol1 == randCol2);

        float temp = inputMatrix(randRow1, randCol1);
        inputMatrix(randRow1, randCol1) = inputMatrix(randRow2, randCol2);
        inputMatrix(randRow2, randCol2) = temp;
    }
    return inputMatrix;
}

auto rbfKernel(Matrix<float, -1, -1> data1, Matrix<float, -1, -1> data2, int sigma) {
    MatrixXf delta = data1 - data2;
    delta = delta.cwiseAbs();
    float squaredEuclidean = delta.sum();
    float result = exp(-squaredEuclidean) / pow(2 * sigma, 2);
    return result;
}

auto thirdTerm(Matrix<float, -1, -1> clusterMember, int var) {
    float result = 0;
    for (int i = 0; i < clusterMember.rows(); i += 1) {
        MatrixXf I = rowMatrix(clusterMember, i);
        for (int j = 0; j < clusterMember.rows(); j += 1) {
            MatrixXf J = rowMatrix(clusterMember, j);
            result = result + rbfKernel(I, J, var);
        }
    }
    result /= pow(clusterMember.rows(), 2);
    return result;
}

float second_term(Matrix<float, -1, -1> dataI, Matrix<float, -1, -1> clusterMember, int var) {
    float result = 0;
    for (int i = 0; i < clusterMember.rows(); i += 1) {
        MatrixXf I = rowMatrix(clusterMember, i);
        result = result + rbfKernel(dataI, I, var);
    }
    result = 2 * result / clusterMember.rows();
    return result;
}

int main() {
    MatrixXf X;
    X.resize(4, 2);
    X << 1, 1,
            2, 3,
            3, 3,
            2.5, 3;
    cout << "X :\n" << X << "\n\n";

    X = shuffle(X);
    cout << "Suffled X :\n" << X << "\n";
}

File: Framework/RadialBasisFunction/rbf_k_means.py
import numpy as np
from matplotlib import pyplot as plt
from matplotlib.pyplot import cm
import time

from test_data import generate_data as grd

# params for test purposes
k = 2  # number of cluster
var = 5  # var in RFB kernel
iteration_counter = 0
data_file = grd.generate_xor(k)  # choice : generate_random, generate_tor, generate_xor, generate_linear
data_testing = np.loadtxt(data_file, delimiter=" ")


# Initiate cluster kernels
def init_cluster(data_input, n_cluster):
    list_cluster_member = [[] for i in range(n_cluster)]
    shuffled_data_in = data_input
    np.random.shuffle(shuffled_data_in)
    for i in range(0, data_input.shape[0]):
        list_cluster_member[i % n_cluster].append(data_input[i, :])

    return list_cluster_member


# New kernel position
def rbf_kernel(data1, data2, sigma):
    delta = abs(np.subtract(data1, data2))
    squared_euclidean = (np.square(delta).sum(axis=1)) # -> Mauvaise traduction en C++ // A revoir
    result = np.exp(-(squared_euclidean) / (2 * sigma ** 2))
    return result


# Compute third term of the mathematical function
def third_term(cluster_member):
    result = 0
    for i in range(0, cluster_member.shape[0]):
        for j in range(0, cluster_member.shape[0]):
            result = result + rbf_kernel(cluster_member[i, :], cluster_member[j, :], var)
    result = result / (cluster_member.shape[0] ** 2)

    return result

# Compute second term of the mathematical function
def second_term(data_I, cluster_member):
    result = 0
    for i in range(0, cluster_member.shape[0]):
        result = result + rbf_kernel(data_I, cluster_member[i, :], var)
    result = 2 * result / cluster_member.shape[0]

    return result


# Display
def plot_result(list_cluster_members, centroid, iteration, converged):
    n = list_cluster_members.__len__()
    color = iter(cm.rainbow(np.linspace(0, 1, n)))
    plt.figure("result")
    plt.clf()
    plt.title("Iteration-" + iteration)
    for i in range(n):
        col = next(color)
        cluster_member = np.asmatrix(list_cluster_members[i])
        plt.scatter(np.ravel(cluster_member[:, 0]), np.ravel(cluster_member[:, 1]), marker=".", s=100, color=col)
    color = iter(cm.rainbow(np.linspace(0, 1, n)))
    for i in range(n):
        col = next(color)
        plt.scatter(np.ravel(centroid[i, 0]), np.ravel(centroid[i, 1]),
                    marker="*", s=400,
                    color=col, edgecolors="black")
    if (converged == 0):
        plt.ion()
        plt.show()
        plt.pause(0.1)
    if (converged == 1):
        plt.show(block=True)


# Main function
def k_means_kernel(data):
    global iteration_counter
    member_init = init_cluster(data, k)
    n_cluster = member_init.__len__()
    # looping until converged
    while (True):
        # Calculate centroid, only for visualization purpose
        centroid = np.ndarray(shape=(0, data.shape[1]))
        for i in range(0, n_cluster):
            cluster_member = np.asmatrix(member_init[i])
            centroid_cluster = cluster_member.mean(axis=0)
            centroid = np.concatenate((centroid, centroid_cluster), axis=0)
        # Plot result in every iteration
        plot_result(member_init, centroid, str(iteration_counter), 0)
        old_time = np.around(time.time(), decimals=0)
        kernel_result_cluster_all_cluster = np.ndarray(shape=(data.shape[0], 0))
        # Assign data to cluster for closest centroid
        for i in range(0, n_cluster):  # repeat for all cluster
            term_3 = third_term(np.asmatrix(member_init[i]))
            matrix_term_3 = np.repeat(term_3, data.shape[0], axis=0);
            matrix_term_3 = np.asmatrix(matrix_term_3)
            matrix_term_2 = np.ndarray(shape=(0, 1))
            for j in range(0, data.shape[0]):  # repeat for all data
                term_2 = second_term(data[j, :], np.asmatrix(member_init[i]))
                matrix_term_2 = np.concatenate((matrix_term_2, term_2), axis=0)
            matrix_term_2 = np.asmatrix(matrix_term_2)
            kernel_result_cluster_I = np.add(-1 * matrix_term_2, matrix_term_3)
            kernel_result_cluster_all_cluster = \
                np.concatenate((kernel_result_cluster_all_cluster, kernel_result_cluster_I), axis=1)
        cluster_matrix = np.ravel(np.argmin(np.matrix(kernel_result_cluster_all_cluster), axis=1))
        list_cluster_member = [[] for l in range(k)]
        for i in range(0, data.shape[0]):  # assign data to cluster regarding cluster matrix
            list_cluster_member[cluster_matrix[i].item()].append(data[i, :])
        for i in range(0, n_cluster):
            print("Cluster member numbers-", i, ": ", list_cluster_member[0].__len__())
        # Break when converged
        bool_acc = True
        for m in range(0, n_cluster):
            prev = np.asmatrix(member_init[m])
            current = np.asmatrix(list_cluster_member[m])
            if (prev.shape[0] != current.shape[0]):
                bool_acc = False
                break
            if (prev.shape[0] == current.shape[0]):
                bool_per_cluster = (prev == current).all()
            bool_acc = bool_acc and bool_per_cluster
            if (bool_acc == False):
                break
        if (bool_acc == True):
            break
        iteration_counter += 1
        # Update new cluster member
        member_init = list_cluster_member
        new_time = np.around(time.time(), decimals=0)
        print("Iteration-", iteration_counter, ": ", new_time - old_time, " seconds")

    return list_cluster_member, centroid


if __name__ == "__main__":
    cluster_result, centroid = k_means_kernel(data_testing)
    plot_result(cluster_result, centroid, str(iteration_counter) + ' (converged)', 1)
    print("Converged !")
